Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,348 @@
+
+    def test_errored_directory(self):
+        """Directory containing wrongly indented python source code files."""
+        with TemporaryDirectory() as tmp_dir:
+            e_file = create_tmp_python_file(
+                SOURCE_CODES["wrong_indented"], dir=tmp_dir
+            )
+            c_file = create_tmp_python_file(SOURCE_CODES["error_free"], dir=tmp_dir)
+            err = ('unindent does not match any outer indentation level'
+                        ' (<tokenize>, line 3)\n')
+            err = "%r: Indentation Error: %s" % (e_file, err)
+            self._verify_tabnanny_check(tmp_dir, err=err)
+
+
+class TestProcessTokens(TestCase):
+    """Testing `tabnanny.process_tokens()`."""
+
+    @patch('tabnanny.NannyNag')
+    def test_with_correct_code(self, MockNannyNag):
+        """A python source code without any whitespace related problems."""
+
+        file_path = create_tmp_python_file(SOURCE_CODES["error_free"])
+        with open(file_path) as f:
+            tabnanny.process_tokens(tokenize.generate_tokens(f.readline))
+        self.assertFalse(MockNannyNag.called)
+
+    def test_with_errored_codes_samples(self):
+        """A python source code with whitespace related sampled problems."""
+
+        # "tab_space_errored_1": executes block under type == tokenize.INDENT
+        #                        at `tabnanny.process_tokens()`.
+        # "tab space_errored_2": executes block under
+        #                        `check_equal and type not in JUNK` condition at
+        #                        `tabnanny.process_tokens()`.
+
+        for key in ["tab_space_errored_1", "tab_space_errored_2"]:
+            file_path = create_tmp_python_file(SOURCE_CODES[key])
+            exception = tabnanny.NannyNag
+            with open(file_path) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+                self.assertRaises(exception, tabnanny.process_tokens, tokens)
+
+class TestCommandLine(TestCase):
+    """Tests command line interface of `tabnanny`."""
+
+    def _validate_cmd(self, *args, return_code=0, stdout=b"", stderr=b""):
+        """Common function to assert the behaviour of command line interface.
+
+        Arguments:
+            * args        : Add arguments which will be directly fed as the
+                            arguments to tabnanny command line.
+            * return_code : Expected return code from tabnanny command line
+                            interface.
+            * stdout      : Expected standard output text after executing
+                            tabnanny command line interface.
+            * stderr      : Expected standard output error text after executing
+                            tabnanny command line interface.
+        """
+        rc, out, err = script_helper.assert_python_ok('-m', 'tabnanny', *args)
+        # Note: The `splitlines()` will solve the problem of CRLF(\r) added
+        # by OS Windows.
```

Comment:
I don't think these chain of methods, `assert_python_ok` -> `_assert_python` -> `run_python_until_end` is allowing to pass `universal_newlines` keyword argument!

Reference: https://github.com/python/cpython/blob/master/Lib/test/support/script_helper.py#L94

Are you suggesting to replace `script_helper.assert_python_ok()` with `subprocess.Popen()` here?